backend/authschemes/localauth/local_auth.go

```diff
@@ -56,39 +74,70 @@ func (LocalAuthScheme) FriendlyName() string {
 // the underlying system/database
 func (p LocalAuthScheme) BindRoutes(r *mux.Router, bridge authschemes.AShirtAuthBridge) {
 	remux.Route(r, "POST", "/register", remux.JSONHandler(func(r *http.Request) (interface{}, error) {
+		if !p.RegistrationEnabled {
+			return nil, fmt.Errorf("registration is closed to users")
+		}
+
 		dr := remux.DissectJSONRequest(r)
-		firstName := dr.FromBody("firstName").Required().AsString()
-		lastName := dr.FromBody("lastName").Required().AsString()
-		email := dr.FromBody("email").Required().AsString()
-		password := dr.FromBody("password").Required().AsString()
+		info := RegistrationInfo{
+			Email:     dr.FromBody("email").Required().AsString(),
+			FirstName: dr.FromBody("firstName").Required().AsString(),
+			LastName:  dr.FromBody("lastName").Required().AsString(),
+			Password:  dr.FromBody("password").Required().AsString(),
+		}
+
 		if dr.Error != nil {
 			return nil, dr.Error
 		}
 
-		if err := checkPasswordComplexity(password); err != nil {
+		if err := checkPasswordComplexity(info.Password); err != nil {
 			return nil, err
 		}
 
-		encryptedPassword, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
-		if err != nil {
-			return nil, backend.WrapError("Unable to generate encrypted password", err)
+		return nil, registerNewUser(r.Context(), bridge, info)
+	}))
+
+	remux.Route(r, "POST", "/admin/register", remux.JSONHandler(func(r *http.Request) (interface{}, error) {
+		if !middleware.IsAdmin(r.Context()) {
+			return nil, backend.UnauthorizedWriteErr(fmt.Errorf("Requesting user is not an admin"))
 		}
 
-		userResult, err := bridge.CreateNewUser(authschemes.UserProfile{
-			FirstName: firstName,
-			LastName:  lastName,
-			Slug:      strings.ToLower(firstName + "." + lastName),
-			Email:     email,
-		})
-		if err != nil {
+		authKey := make([]byte, 42)
+		if _, err := rand.Read(authKey); err != nil {
+			return nil, backend.WrapError("Unable to generate random new user password key", err)
+		}
+
+		// convert authKey into readable format
+		bufKey := make([]byte, len(authKey))
+		for i, b := range authKey {
```

Comment:
If we're using base64 we don't need to do the modulo math here to get into an ascii printable space because base64 is always printable. This still reduces the keyspace because the input going into base64 is constrained to just the printable space already.

Reply:
Oh, sorry, i meant to remove that. I guess I was too eager with my commit.